logstash-core/src/main/java/org/logstash/execution/AbstractPipelineExt.java

```diff
@@ -364,6 +362,35 @@ public final JRubyWrappedWriteClientExt wrappedWriteClient(final ThreadContext c
             .initialize(inputQueueClient, pipelineId.asJavaString(), metric, pluginId);
     }
 
+    @JRubyMethod(name = "pipeline_source_details", visibility = Visibility.PROTECTED)
+    @SuppressWarnings("rawtypes")
+    public RubyArray getPipelineSourceDetails(final ThreadContext context) {
+        RubyArray res = (RubyArray) pipelineSettings.callMethod(context, "config_parts");
+        List<RubyString> pipelineSources = new ArrayList<>(res.size());
+        for (IRubyObject part : res.toJavaArray()) {
+            SourceWithMetadata sourceWithMetadata = part.toJava(SourceWithMetadata.class);
+            String protocol = sourceWithMetadata.getProtocol();
+            switch (protocol) {
+                case "string":
+                    pipelineSources.add(RubyString.newString(context.runtime, "config string"));
+                    break;
+                case "file":
+                    pipelineSources.add(RubyString.newString(context.runtime, sourceWithMetadata.getId()));
+                    break;
+                case "x-pack-metrics":
+                    pipelineSources.add(RubyString.newString(context.runtime, "monitoring pipeline"));
+                    break;
+                case "x-pack-config-management":
+                    pipelineSources.add(RubyString.newString(context.runtime, "central pipeline management"));
+                    break;
+                case "module":
```

Comment:
Ideally I think we might want the name of the module here, but it might not be too important as we should be able to glean it from elsewhere